framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -1285,10 +1287,123 @@ protected void postDirectSuperTypes(
      */
     public void postAsMemberOf(
             AnnotatedTypeMirror type, AnnotatedTypeMirror owner, Element element) {
+        if (element.getKind() == ElementKind.FIELD) {
+            addAnnotationFromFieldInvariant(type, owner, (VariableElement) element);
+        }
         addComputedTypeAnnotations(element, type);
     }
 
     /**
+     * Adds the qualifier specified by a field invariant for {@code field} to {@code type}
+     *
+     * @param type annotated type to which the annotation is added
+     * @param accessedVia the annotated type of the receiver of the accessing tree. (Only used to
+     *     get the type element of the underling type.)
+     * @param field element of the field
+     */
+    protected void addAnnotationFromFieldInvariant(
+            AnnotatedTypeMirror type, AnnotatedTypeMirror accessedVia, VariableElement field) {
+        TypeElement typeElement = InternalUtils.getTypeElement(accessedVia.getUnderlyingType());
+        if (ElementUtils.enclosingClass(field).equals(typeElement)) {
+            // If the field is declared in the accessedVia class, then the field in the invariant
+            // cannot be this field, even if the field has the same name.
+            return;
+        }
+
+        FieldInvariantObject invariants = getFieldInvariants(typeElement);
+        if (invariants == null) {
+            return;
+        }
+        List<AnnotationMirror> invariantAnnos = invariants.getQualifiersFor(field.getSimpleName());
+        type.replaceAnnotations(invariantAnnos);
+    }
+
+    /**
+     * Returns the field invariants for the given class.
+     *
+     * <p>Subclass may override this to implement their own field invariant annotations if {@link
```

Comment:
Really, subclasses may implement their own field invariant annotations, and in that case they need to override this too.